### PR TITLE
Plane: abstract flight modes into new Mode classes

### PR DESCRIPTION
**Pull request by magicrub**
Plane: abstract flight modes into new Mode classes. This brings us closer to the mode layout of Copter and Rover. This only effects set_mode() so far, it's a mergable baby step toward a long-term solution.

### CODE REVIEW

File: ArduPlane/ArduPlane.cpp [hunk 1]
```diff
@@ -690,7 +690,7 @@ void Plane::update_navigation()
         qrtl_radius = abs(aparm.loiter_radius);
     }
     
-    switch(control_mode) {
+    switch ((FlightMode)control_mode->mode_number()) {
```
Comment (peterbarker) [inline ArduPlane/ArduPlane.cpp]: Why is the cast necessary?

File: ArduPlane/GCS_Mavlink.cpp [hunk 2]
```diff
@@ -87,12 +87,12 @@ MAV_MODE GCS_MAVLINK_Plane::base_mode() const
 
 uint32_t GCS_MAVLINK_Plane::custom_mode() const
 {
-    return plane.control_mode;
+    return plane.control_mode->mode_number();
```
Comment (peterbarker) [inline ArduPlane/GCS_Mavlink.cpp]: A cast here will be required when the return type from `mode_number()` is corrected.

File: ArduPlane/GCS_Mavlink.cpp [hunk 3]
```diff
@@ -1686,28 +1686,9 @@ AP_Rally *GCS_MAVLINK_Plane::get_rally() const
  */
 bool GCS_MAVLINK_Plane::set_mode(const uint8_t mode)
```
Comment (peterbarker) [inline ArduPlane/GCS_Mavlink.cpp]: It's odd that we send a `uint32_t` down to GCS_MAVLINK but take a `uint8_t` back up.  All vehicles will probably do this ATM.

File: ArduPlane/Plane.h [hunk 4]
```diff
@@ -1022,7 +1065,7 @@ class Plane : public AP_HAL::HAL::Callbacks {
     void do_digicam_control(const AP_Mission::Mission_Command& cmd);
     bool start_command_callback(const AP_Mission::Mission_Command &cmd);
     bool verify_command_callback(const AP_Mission::Mission_Command& cmd);
-    void notify_flight_mode(enum FlightMode mode);
+    void notify_mode(const Mode *mode);
```
Comment (peterbarker) [inline ArduPlane/Plane.h]: The argument here should never be null; pass a reference instead.

File: ArduPlane/control_modes.cpp [hunk 5]
```diff
@@ -1,5 +1,75 @@
...
+    case Mode::Number::LOITER:
+        ret = &mode_loiter;
+        break;
+    case Mode::Number::AVOID_ADSB:
+        ret = &mode_avoidADSB;
+        break;
+    case Mode::Number::GUIDED:
+        ret = &mode_guided;
+        break;
+    case Mode::Number::INITIALISING:
```
Comment (peterbarker) [inline ArduPlane/control_modes.cpp]: Love the spelling here.  I think we typically canonicalise on American English.

File: ArduPlane/events.cpp [hunk 6]
```diff
@@ -16,34 +16,34 @@ void Plane::failsafe_short_on_event(enum failsafe_state fstype, mode_reason_t re
...
         break;
         
     case AUTO:
     case AVOID_ADSB:
     case GUIDED:
     case LOITER:
         if(g.fs_action_short != FS_ACTION_SHORT_BESTGUESS) {
-            failsafe.saved_mode = control_mode;
+            failsafe.saved_mode = (FlightMode)control_mode->mode_number();
             failsafe.saved_mode_set = true;
```
Comment (peterbarker) [inline ArduPlane/events.cpp]: Getting rid of this and making `failsafe.saved_mode` ought to be high on the to-do-next list.

File: ArduPlane/events.cpp [hunk 7]
```diff
@@ -55,7 +55,7 @@ void Plane::failsafe_short_on_event(enum failsafe_state fstype, mode_reason_t re
     default:
         break;
     }
-    gcs().send_text(MAV_SEVERITY_INFO, "Flight mode = %u", (unsigned)control_mode);
+    gcs().send_text(MAV_SEVERITY_INFO, "Flight mode = %u", (unsigned)control_mode->mode_number());
```
Comment (peterbarker) [inline ArduPlane/events.cpp]: `%s` and `->name4()` would be awesome here :-)

File: ArduPlane/geofence.cpp [hunk 8]
```diff
@@ -291,14 +291,14 @@ void Plane::geofence_check(bool altitude_check_only)
...
-            (control_mode == GUIDED || control_mode == AVOID_ADSB) &&
+            (control_mode == &mode_guided || control_mode == &mode_avoidADSB) &&
             geofence_present() &&
             geofence_state->boundary_uptodate &&
             geofence_state->old_switch_position == oldSwitchPosition &&
             guided_WP_loc.lat == geofence_state->guided_lat &&
             guided_WP_loc.lng == geofence_state->guided_lng) {
             geofence_state->old_switch_position = 254;
-            set_mode(get_previous_mode(), MODE_REASON_GCS_COMMAND);
+            set_mode(*previous_mode, MODE_REASON_GCS_COMMAND);
```
Comment (peterbarker) [inline ArduPlane/geofence.cpp]: Some sanity checks here for `previous_mode` not being `nullptr`.  Best to be safe.

File: ArduPlane/mode.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+void Mode::exit()
+{
+    // call sub-classes exit
+    _exit();
+}
+
+bool Mode::enter()
+{
+    // cancel inverted flight
+     plane.auto_state.inverted_flight = false;
```
Comment (peterbarker) [inline ArduPlane/mode.cpp]: You are nuking vast amounts of state here.  If the subclass enter fails - what happens?  I *believe* Plane doesn't let that happen - but this is a landmine waiting to happen.

File: ArduPlane/mode.h [hunk 10]
```diff
@@ -0,0 +1,475 @@
...
+#include <stdint.h>
+
+#include <GCS_MAVLink/GCS_MAVLink.h>
+#include <AP_Math/AP_Math.h>
+
+#include "defines.h"
+
+#define MODE_NEXT_HEADING_UNKNOWN   99999.0f    // used to indicate to set_desired_location method that next leg's heading is unknown
+
+// pre-define ModeRTL so Auto can appear higher in this file
```
Comment (peterbarker) [inline ArduPlane/mode.h]: Is this required?

File: ArduPlane/mode.h [hunk 11]
```diff
@@ -0,0 +1,475 @@
+#pragma once
+
+#include <stdint.h>
+
+#include <GCS_MAVLink/GCS_MAVLink.h>
+#include <AP_Math/AP_Math.h>
+
+#include "defines.h"
+
+#define MODE_NEXT_HEADING_UNKNOWN   99999.0f    // used to indicate to set_desired_location method that next leg's heading is unknown
```
Comment (peterbarker) [inline ArduPlane/mode.h]: This is definitely not required.

File: ArduPlane/mode.h [hunk 12]
```diff
@@ -0,0 +1,475 @@
...
+        QHOVER        = 18,
+        QLOITER       = 19,
+        QLAND         = 20,
+        QRTL          = 21
+    };
+
+    // Constructor
+    Mode();
+
+    // enter this mode, returns false if we failed to enter
```
Comment (peterbarker) [inline ArduPlane/mode.h]: Plane doesn't let you fail to enter a mode, does it?  That would make the comment misleading.

File: ArduPlane/mode.h [hunk 13]
```diff
@@ -0,0 +1,475 @@
...
+    //
+    // navigation methods
+    //
+
+    // return distance (in meters) to destination
+    virtual float get_distance_to_destination() const { return 0.0f; }
+
+protected:
+
+    // subclasses override this to perform checks before entering the mode
```
Comment (peterbarker) [inline ArduPlane/mode.h]: Then why isn't it `const`?

File: ArduPlane/mode.h [hunk 14]
```diff
@@ -0,0 +1,475 @@
...
+
+
+class ModeAcro : public Mode
+{
+public:
+    uint32_t mode_number() const override { return Number::ACRO; }
+    const char *name4() const override { return "ACRO"; }
+
+    // methods that affect movement of the vehicle in this mode
+    void update() override { }
```
Comment (peterbarker) [inline ArduPlane/mode.h]: This empty implementation should probably go in the base class.

File: ArduPlane/mode.h [hunk 15]
```diff
@@ -0,0 +1,475 @@
...
+
+    uint32_t mode_number() const override { return Number::AUTO; }
+    const char *name4() const override { return "AUTO"; }
+
+    // methods that affect movement of the vehicle in this mode
+    void update() override { }
+
+    // attributes for mavlink system status reporting
+    bool has_manual_input() const override { return true; }
+    bool attitude_stabilized() const override { return false; }
```
Comment (peterbarker) [inline ArduPlane/mode.h]: Really?

File: ArduPlane/mode.h [hunk 16]
```diff
@@ -0,0 +1,475 @@
...
+public:
+
+    uint32_t mode_number() const override { return Number::INITIALISING; }
+    const char *name4() const override { return "INIT"; }
+
+    // methods that affect movement of the vehicle in this mode
+    void update() override { }
+
+    // attributes for mavlink system status reporting
+    bool has_manual_input() const override { return true; }
```
Comment (peterbarker) [inline ArduPlane/mode.h]: Yeah?

File: ArduPlane/mode_qland.cpp [hunk 17]
```diff
@@ -0,0 +1,17 @@
+#include "mode.h"
+#include "Plane.h"
+
+bool ModeQLand::_enter()
+{
+    plane.throttle_allows_nudging = true;
+    plane.auto_navigation_mode = false;
+    if (!plane.quadplane.init_mode()) {
+        plane.control_mode = plane.previous_mode;
```
Comment (peterbarker) [inline ArduPlane/mode_qland.cpp]: Didn't we nuke a whole lot of state before getting here?

File: ArduPlane/system.cpp [hunk 18]
```diff
@@ -263,249 +266,48 @@ void Plane::startup_ground(void)
...
-    // perform any cleanup required for prev flight mode
-    exit_mode(control_mode);
-
-    // cancel inverted flight
-    auto_state.inverted_flight = false;
-
-    // don't cross-track when starting a mission
-    auto_state.next_wp_crosstrack = false;
+    // backup current control_mode and previous_mode
+    Mode &old_previous_mode = *previous_mode;
```
Comment (peterbarker) [inline ArduPlane/system.cpp]: Sanity check?

File: ArduPlane/mode.h (hunk 14 above)
Comment (magicrub) [reply ArduPlane/mode.h]: so, remove the pure virtual?

File: ArduPlane/system.cpp (hunk 18 above)
Comment (magicrub) [reply ArduPlane/system.cpp]: fixed

File: ArduPlane/mode.h (hunk 14 above)
Comment (magicrub) [reply ArduPlane/mode.h]: Let's save it for now

File: ArduPlane/Plane.h (hunk 4 above)
Comment (magicrub) [reply ArduPlane/Plane.h]: fixed

File: ArduPlane/GCS_Mavlink.cpp (hunk 3 above)
Comment (magicrub) [reply ArduPlane/GCS_Mavlink.cpp]: works as-is

File: ArduPlane/GCS_Mavlink.cpp (hunk 2 above)
Comment (magicrub) [reply ArduPlane/GCS_Mavlink.cpp]: fixed

File: ArduPlane/ArduPlane.cpp (hunk 1 above)
Comment (magicrub) [reply ArduPlane/ArduPlane.cpp]: fixed, removed FlightMode

**Comment by magicrub** [comment]
migrated a few more things to mode::

File: ArduPlane/events.cpp (hunk 7 above)
Comment (magicrub) [reply ArduPlane/events.cpp]: sure would! After this PR, and a dozen other related PRs, I'd love to come back and do that!

File: ArduPlane/geofence.cpp (hunk 8 above)
Comment (magicrub) [reply ArduPlane/geofence.cpp]: I've added this check, although I think previous_mode will get removed. Hopefully removed. It's kind of a hack the way it is now.

File: ArduPlane/mode.cpp (hunk 9 above)
Comment (magicrub) [reply ArduPlane/mode.cpp]: yes, it has always been a landmine. If we change things to allow set_mode to ever fail then a whole redesign will be needed.

File: ArduPlane/mode.h (hunk 13 above)
Comment (magicrub) [reply ArduPlane/mode.h]: huh?

File: ArduPlane/mode.h (hunk 11 above)
Comment (magicrub) [reply ArduPlane/mode.h]: fixed. I removed a bunch of the copy/paste leftover stuff at the top

File: ArduPlane/mode.h (hunk 10 above)
Comment (magicrub) [reply ArduPlane/mode.h]: nope, fixed

File: ArduPlane/mode.h (hunk 12 above)
Comment (magicrub) [reply ArduPlane/mode.h]: fixed

File: ArduPlane/mode.h (hunk 15 above)
Comment (magicrub) [reply ArduPlane/mode.h]: fixed

File: ArduPlane/mode.h (hunk 16 above)
Comment (magicrub) [reply ArduPlane/mode.h]: fixed

**Comment by peterbarker** [comment]
Factoring out a qenter and qupdate saved 1.3kB

**Comment by magicrub** [comment]
Same can be done with avoid and guided, update too. All three functions including update can be done on q as well.

**Comment by magicrub** [comment]
@peterbarker do you have a commit for that qenter consolidation? Go ahead and push it

**Comment by magicrub** [comment]
factoring out q_enable() & q_update() saves 1272 bytes
factoring out guided_enable() & guided_update() saves 104 bytes
factoring out fbwa_update() saves 292 bytes

total saving: 1668

**Pull request closed without merging**